This PR contains updates to the Terraform modules

paolosalvatori opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
This PR contains updates to the Terraform modules:

- Introduced variables for all the properties of Azure resources
- Extended AKS clusters and node pools with new options and properties
- Introduced the following resources:
  - Azure Bastion
  - Azure Private DNS Zones
  - Azure Private Endpoints
  - Azure Diagnostics Settings
  - Azure Log Analytics
  - Azure Container Registry with two geo-replications
  - User-Assigned Managed Identities

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
terraform init
terraform plan
terraform apply
```

* Test the code
<!-- Add steps to run the tests suite and/or manually test -->
```
```

## What to Check
Verify that the following are valid
* ...

## Other Information
<!-- Add any other helpful information that may be needed here. -->

paolosalvatori commented on istio/log_analytics.tf:
```diff
@@ -0,0 +1,29 @@
+resource "azurerm_log_analytics_workspace" "log_analytics_workspace_one" {
```
> This creates 2 distinct workspaces. Why not create a single one so that all logs are in the same workspace ? In this sample there is a single multi-cluster service mesh, so it makes sense to have all logs in 1 place. What do you think ?

By default, creating a single workspace is wrong or at least suboptimal. Writing monitoring data across regions can generate additional costs, especially when the source resources are in a different region with respect to the workspace.

zioproto commented on istio/main.tf:
```diff
@@ -2,109 +2,349 @@ resource "random_string" "random" {
-  sku_tier                          = "Paid" # defaults to Free
-  role_based_access_control_enabled = true
+module "aks_one" {
+  source                            = "Azure/aks/azurerm"
```
Why in commit dc07af46e907127dddfb8b407ba3871a3c55c411 you dropped this line?
```
version                           = "6.0.0"
```

paolosalvatori replied:
I dropped the line because it was causing issues. I won't include it again.

zioproto commented on istio/main.tf:
```diff
@@ -2,109 +2,349 @@ resource "random_string" "random" {
-  sku_tier                          = "Paid" # defaults to Free
-  role_based_access_control_enabled = true
+module "aks_two" {
+  source                            = "Azure/aks/azurerm"
```
Again here the `version` is lost

paolosalvatori replied:
Again, I won't include the version as it's causing issues.

zioproto commented on istio/variables.tf:
```diff
@@ -1,18 +1,20 @@
+  description = "The location of the second cluster and related resources."
   type        = string
-  default     = null
+  default     = "eastus"
```
in the file `terraform.tfvars` you have a different default value

paolosalvatori replied:
That is correct: in variables.tf you specify default values while in terraform.tfvars you specify actual values

paolosalvatori replied:
Said that I can even specify a default value for location_one and location_two, but it's purely indicative. When publishing ARM templates to Azure Quickstart Templates I was required not to hardcode default values for variables such as locations as a rule of thumb.

paolosalvatori replied:
You tried a scenario across with two regions across two geographies, I tried a scenario with two paired regions within the same geography.

zioproto commented on istio/terraform.tfvars:
```diff
@@ -0,0 +1,65 @@
+kubernetes_version                  = "1.23.5"
+orchestrator_version                = "1.23.5"
+location_one                        = "westeurope"
+location_two                        = "northeurope"
```
in `istio/variables.tf` you have a different default value for `location_two`

paolosalvatori replied:
Again, this is intentional: in variables.tf you specify default values while in terraform.tfvars you specify actual values

zioproto commented on istio/virtual_network.tf:
```diff
@@ -0,0 +1,100 @@
+module "network_one" {
+  source              = "Azure/network/azurerm"
```
I know this was missing also in the original version, but can we add the version of the module ?
```
module "network" {
  source  = "Azure/network/azurerm"
  version = "3.5.0"
}
```

paolosalvatori replied:
No, I prefer not. There is a warning in the current version. If I hardcode the version, the module will never get the advantage of fixes.

zioproto commented on istio/virtual_network.tf:
```diff
@@ -0,0 +1,100 @@
+}
+
+module "network_two" {
+  source              = "Azure/network/azurerm"
```
I know this was missing also in the original version, but can we add the version of the module ?
```
module "network" {
  source  = "Azure/network/azurerm"
  version = "3.5.0"
}
```

paolosalvatori replied:
If you hardcode the version, the project will never get the advantage of fixes to the Azure/network/azurerm module. There's a warning now.

zioproto commented on istio/variables.tf:
```diff
@@ -1,18 +1,20 @@
-variable "resource_group_name" {
-  description = "The resource group name to be imported"
+variable "name_prefix" {
```
This file has 76 variables. Do we really need to have all of them in this file, or we can remove some when the default value is in the AKS module ?

paolosalvatori replied:
The 76 variables are correct, the vast majority have a default value, so you don't necessarily have to touch them. If you want to use a different value instead, you don't have to change the form. Do you think hardcoding the values is correct? Besides being an SRE, have you ever participated in a real project as a consultant or something else? Hardcoding values is an anti-pattern.

zioproto commented on istio/variables.tf:
```diff
@@ -1,18 +1,20 @@
+  default     = "eastus"
 }
 
 variable "client_id" {
```
Is `client_id` ever used ?

paolosalvatori replied:
We can remove this variable, I found it in the original variables.tf, so you added this variable, not me 😸 I removed it.

zioproto commented on istio/variables.tf:
```diff
@@ -153,169 +138,345 @@ variable "rbac_aad_server_app_secret" {
+  default     = []
+}
+
+variable "enable_node_public_ip" {
```
Can we drop this one ? We dont use this variable `enable_node_public_ip`

paolosalvatori replied:
Same as above, I found enable_node_public_ip in variables.tf so you added this variable. 😄 I removed it.

zioproto commented on istio/variables.tf:
```diff
@@ -153,169 +138,345 @@ variable "rbac_aad_server_app_secret" {
+  default     = "user"
+}
+
+variable "user_agents_labels" {
```
Also this one we are not using `user_agents_labels`

paolosalvatori replied:
Actually, we are using `user_agents_labels` in `node_pool.tf`

zioproto commented on istio/variables.tf:
```diff
@@ -153,169 +138,345 @@ variable "rbac_aad_server_app_secret" {
 
-variable "orchestrator_version" {
-  description = "Specify which Kubernetes release to use for the orchestration layer. The default used is the latest Kubernetes version available in the region"
+variable "node_resource_group" {
```
Also `node_resource_group` is not used

paolosalvatori replied:
node_resource_group is another variable I found in variables.tf. I ask you the question: why didn't you delete the variables not used by variables.tf at the beginning? Were you in too much of a hurry? I understand it was a quick and dirty project in the beginning, but now it is public, you should have done it yourself the prune of the unused variables. I just added the missing ones to remove hardcoded values.

zioproto reviewed (changes requested):
You need to update `clusterName` in `002-multicluster-region-one.yaml` and `002-multicluster-region-two.yaml`

zioproto commented on istio/terraform.tfvars:
```diff
@@ -0,0 +1,65 @@
+  servicemesh = "Istio"
+}
+acr_sku                               = "Premium"
+log_analytics_workspace_sku           = "PerNode"
```
I think here you need `PerGB2018` that is the default value, and you cannot use anymore `PerNode`.

See https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/log_analytics_workspace

![Screenshot 2022-09-19 at 21 46 53](https://user-images.githubusercontent.com/789701/191103381-00510d89-3796-4663-bf06-b4f772944dfe.png)

paolosalvatori replied:
I usually use `PerGB2018` so all good, I used `PerNode` because the default was `Free` and `PerNode` was the first that came to my mind while testing 😄 I changed the value to `PerGB2018 `. Note: this is the default value for log_analytics_workspace_sku in `variables.tf`

paolosalvatori commented on istio/README.md:
````diff
@@ -1,188 +1,230 @@
 
-```
+```sh
 cp  tfvars .tfvars
````
The file was not deleted, it was renamed to terraform.tfvars that is the standard name for tfvars files according to Terraform documentation. I fixed it.

zioproto commented on istio/README.md:
```diff
@@ -1,188 +1,270 @@
+  - `System` used by the AKS system node pool
+  - `User` used by the AKS user node pool
+  - `AppGateway` used by the Application Gateway
+  - `AzureBastionHost` used bu Azure Bastion
```
bu -> by

paolosalvatori replied:
fixed

zioproto commented on istio/README.md:
````diff
@@ -1,188 +1,270 @@
-vim .tfvars #customize if needed
-terraform init -upgrade && terraform apply -var-file=.tfvars
+```sh
+cp  terraform .tfvars
````
This `cp` line you can remove completely

paolosalvatori replied:
removed

zioproto commented on istio/README.md:
````diff
@@ -1,188 +1,270 @@
-kubectl --context=eastus-aks get nodes
-kubectl --context=westeurope-aks get nodes
+```sh
+aksClusterOneName="<aks-cluster-one-name>"
````
This step is done already at lines 51-52

paolosalvatori replied:
I changed scripts so this is no more relevant, we need to recheck when finished the README

zioproto commented on istio/README.md:
````diff
@@ -1,188 +1,270 @@
+```sh
+sharedResourceGroupName="..."
+
+export keyVaultName=$(az keyvault list --resource-group $sharedResourceGroupName --query [].name --output tsv)
````
This is not working on `zsh`:
```
 % az keyvault list --resource-group $sharedResourceGroupName --query [].name --output tsv
zsh: no matches found: [].name
```

zioproto replied:
You can make it portable to `zsh` using double quotes around the query:

`"[0].name"`

paolosalvatori commented:
- I changed the name of the istio-installation folder to yaml because in the same directory I clone the Istio repo otherwise I have two folders called istio and istio-installation
- I added the scripts in the scripts directory. We need to add the missing scripts
- I have added comments to the scripts, so far I have not added various controls
- I removed the secret_provider_class.tf file and moved the outputs to outputs.tf
- Fixed * .tftpl files (check them)

zioproto commented:
I added commit https://github.com/Azure-Samples/aks-multi-cluster-service-mesh/pull/4/commits/1389749feb17529e0af9a3e74dc1652653e589fb to fix a small problem in a script, when calling `kubectl` I was pointing to the wrong cluster.

The PR LGTM ! I propose to merge and work on future improvements in smaller PRs that will be easier to review.

Thank you !!!!
